core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -386,6 +390,15 @@ private void completeHeaders(Metadata.Headers headers) {
       headers.put(TIMEOUT_KEY, method.getTimeout());
     }
 
+    private void startDeadlineTimer() {
+      SharedResourceHolder.get(TIMER_SERVICE).schedule(new Runnable() {
```

Comment:
We should probably hang onto the returned `ScheduledFuture` and cancel it `ClientStreamListenerImpl.closed` when we set `closed = true`. That way we won't retain many old streams in memory if they had a long deadline. It looks like it will need to be a volatile reference, though.